cgrok/tree.c

```diff
@@ -0,0 +1,275 @@
+// keys and values, so we can't depend on null-termination
+// to find the lenght of a key.
+void *tcpack(const void *value, uint32_t size) {
+  char *buf = malloc(size + 5);
+  if (!buf) {
+    return NULL;
+  }
+  memcpy(buf, &size, 4);
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
+  return buf;
+}
+
+// Insert a key-value pair. If the key already exists the value will be overwritten
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    
+  }
+  void *val = tcpack(vbuf, vsiz);
+  if (!val) {
+
+  }
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
+  if (!inserted) {
+    free(key);
+    free(*valPtr);
+  }
+  *valPtr = val;
+}
+
+// Insert a key-value pair. If the key already exists return false and keep the original value
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    fprintf(stderr, "Failed to malloc tree key\n");
+    return;
+  } 
+  void *val = tcpack(vbuf, vsiz); 
+  if (!val) {
+    fprintf(stderr, "Failed to malloc tree value\n");
+    return;
+  }
+  bool inserted;
+  void **valPtr = dict_insert(tree->dict, key, &inserted);
+  if (inserted) {
+    *valPtr = val;
+  } else {
+    free(key);
+    free(val); 
+  }
+  return inserted;
+}
+
+// Get the value for the given key, or NULL if the key is not in the tree
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp) {
+  const void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    fprintf(stderr, "Failed to malloc tree key\n");
+    return;
+  }
+  void *value = dict_search(tree->dict, key);
+  *sp = 0;
+  free(key);
+  if (value) {
+    *sp = *(uint32_t*)value;
+    return value+4;
+  }
+  return NULL;
+}
+
+// Remove all elements from the tree
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+// Free the tree and associated iterator
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
+TCLIST *tclistnew(void) {
+  TCLIST *list = malloc(sizeof(TCLIST));
+  if (!list) {
+    return NULL;
+  }
+  list->len = 0;
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  if (!newElem) {
+    return NULL;
+  }
+  newElem->elem = 0;
+  list->head = newElem;
+  return list;
+}
+
+// Get the number of elements in the list
+int tclistnum(const TCLIST *list) {
+  return list->len;
+}
+
+// Null-terminate the value to be inserted, as TokyoCabinet does
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
+  if (!elem) {
+    return NULL;
+  }
+  memcpy(elem, buf, size);
+  ((char *)elem)[size] = 0;
+  return elem;
+}
+
+// Append a new element at the end of the list
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
```

Comment:
nit: This variable is only the `head` of the list for a short while. Something like `item` or `cur` would be better.